### PR TITLE
mask random seed fix

### PR DESCRIPTION
**Pull request by z-fabian**
When using `ddp` training, workers with the same `id` across different  devices share the same initial random seed for the mask function. This decreases the diversity of random masks during training, since the same mask will be generated on different devices throughout training. This fix generates a unique initial random seed for each worker based on their `id`, device `rank` and dataset (in case of `combine_train_val = True`).

### CODE REVIEW

File: fastmri/pl_modules/data_module.py [hunk 1]
```diff
@@ -21,14 +21,38 @@ def worker_init_fn(worker_id):
...
         SliceDataset, CombinedSliceDataset
     ] = worker_info.dataset  # pylint: disable=no-member
+
+    # Check if we are using DDP
+    is_ddp = False
+    if torch.distributed.is_available():
+        if torch.distributed.is_initialized():
+            is_ddp = True
+            device_rank = torch.distributed.get_rank()
+            num_gpus = torch.distributed.get_world_size()
```
Comment (mmuckley) [inline fastmri/pl_modules/data_module.py:31]: No need for this, just use `torch.distributed.get_rank()`.

File: fastmri/pl_modules/data_module.py [hunk 2]
```diff
@@ -21,14 +21,38 @@ def worker_init_fn(worker_id):
...
+            num_workers = worker_info.num_workers
```
Comment (mmuckley) [inline fastmri/pl_modules/data_module.py:32]: Already handled by `worker_info`.

**Review by mmuckley — CHANGES_REQUESTED** [review]
There's some question if this is expected behavior when using `pl_seed_everything()`. Based on PyTorch Lightning issues, this might be working normally. For me, I think it is reasonable to expect different random seeds for each GPU even when seeding everything, so it is okay to move forward with this.

You don't need to worry about the workers. These are handled by `torch.utils.data.DataLoader` via the `worker_info` variable. So we only need to worry about incrementing the seed based on global rank. `torch.distributed.get_rank()` returns the global rank, so the final seed in the distributed setting should be

```python
seed = (worker_info.seed + torch.distributed.get_rank() * worker_info.num_workers) % (2 ** 32 - 1)
```

and I think the non-distributed setting can remain as-is. Let me know if I made any mistakes here.

**Comment by z-fabian** [comment]
@mmuckley I agree that this is probably expected behavior when using `pl_seed_everything()` and it works normally, but I think it makes sense to generate completely unique masks in each worker (even though the data it is applied to will be different). There is even a slight improvement in validation ssim on smaller subsampled datasets. 

Also I agree that `seed = (worker_info.seed + torch.distributed.get_rank() * worker_info.num_workers) % (2 ** 32 - 1)` should be used in the distributed case when there is a single dataset. In the non-distributed setting we should fall back on the same seed as before. 

I included this part
```
        if torch.distributed.is_initialized():
            is_ddp = True
            device_rank = torch.distributed.get_rank()
            num_gpus = torch.distributed.get_world_size()`  
```
to check whether distributed training is being used. If not, we probably cannot call `torch.distributed.get_rank()`. Is there a simpler way of doing this? I might have misunderstood your comment on this part.

**Comment by mmuckley** [comment]
@z-fabian I thought your original code with a serial check of `torch.distributed.is_available()` and then `torch.distributed.is_initialized()` was reasonable. After this we should be able to call `torch.distributed.get_rank()` and get the global rank, unless I'm missing something. We shouldn't need the GPU count - we just need our process rank.

**Comment by z-fabian** [comment]
@mmuckley Oh, I see. I use the GPU count to generate unique seed when we have multiple datasets here: 
```
                    seed_i = (
                        base_seed
                        + device_rank * num_workers
                        + i * num_gpus * num_workers
                    ) % (2 ** 32 - 1)
```
because here we have 3 'dimensions' that can vary. That is we need a unique seed for each dataset on each worker on each GPU. But I could base it on the total number of datasets if that's preferable, like this: 
```
                    seed_i = (
                        base_seed
                        + i * num_workers
                        + device_rank * num_datasets * num_workers
                    ) % (2 ** 32 - 1)
```
Am I missing something here? Is there anything else I should change?

**Comment by mmuckley** [comment]
Oh I see, in that case the "spread" for each GPU process is `num_workers * num_datasets`. In that case, For a single dataset we need

```python
seed = base_seed + torch.distributed.get_rank() * worker_info.num_workers
```

For single dataset code on a 2-GPU problem with 4 workers per GPU this gives:

```bash
rank: 1, worker: 0, seed: 734796318
rank: 0, worker: 1, seed: 734796315
rank: 0, worker: 2, seed: 734796316
rank: 1, worker: 1, seed: 734796319
rank: 0, worker: 3, seed: 734796317
rank: 1, worker: 2, seed: 734796320
rank: 1, worker: 3, seed: 734796321
```

For multiple datasets, we need:

```python
seed_i = (
    base_seed
    - worker_info.id
    + torch.distributed.get_rank()
    * (worker_info.num_workers * len(data.datasets))
    + worker_info.id * len(data.datasets)
    + i
) % (2 ** 32 - 1)
```

This multiple dataset code on a 2-GPU problem with 4 workers per GPU gives:

```bash
rank: 0, worker: 0, dataset: 0, seed: 2723649642
rank: 0, worker: 0, dataset: 1, seed: 2723649643
rank: 1, worker: 0, dataset: 0, seed: 2723649650
rank: 1, worker: 0, dataset: 1, seed: 2723649651
rank: 0, worker: 1, dataset: 0, seed: 2723649644
rank: 0, worker: 1, dataset: 1, seed: 2723649645
rank: 1, worker: 1, dataset: 0, seed: 2723649652
rank: 1, worker: 1, dataset: 1, seed: 2723649653
rank: 0, worker: 2, dataset: 0, seed: 2723649646
rank: 0, worker: 2, dataset: 1, seed: 2723649647
rank: 1, worker: 2, dataset: 0, seed: 2723649654
rank: 1, worker: 2, dataset: 1, seed: 2723649655
rank: 0, worker: 3, dataset: 0, seed: 2723649648
rank: 0, worker: 3, dataset: 1, seed: 2723649649
rank: 1, worker: 3, dataset: 0, seed: 2723649656
rank: 1, worker: 3, dataset: 1, seed: 2723649657
```

**Ignore this**

The below code is just a reference for how the default seeding works. On a 2-GPU DDP job with 4 workers per GPU, we get:

```bash
torch.distributed.get_rank(): 1, worker.info.seed: 8314211556539077902
torch.distributed.get_rank(): 0, worker.info.seed: 8314211556539077902
torch.distributed.get_rank(): 1, worker.info.seed: 8314211556539077903
torch.distributed.get_rank(): 0, worker.info.seed: 8314211556539077903
torch.distributed.get_rank(): 1, worker.info.seed: 8314211556539077904
torch.distributed.get_rank(): 0, worker.info.seed: 8314211556539077904
torch.distributed.get_rank(): 1, worker.info.seed: 8314211556539077905
torch.distributed.get_rank(): 0, worker.info.seed: 8314211556539077905
```

For a 4-node, 32-GPU job, on the 5th GPU on the 3rd node with 10 workers I get:

```bash
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313846
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313847
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313848
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313849
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313850
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313851
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313852
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313853
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313854
torch.distributed.get_rank(): 20, worker.info.seed: 6298923453764313855
```

The 2nd GPU on the 1st node is:

```bash
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313846
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313847
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313848
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313849
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313850
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313851
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313852
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313853
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313854
torch.distributed.get_rank(): 1, worker.info.seed: 6298923453764313855
```

**Comment by z-fabian** [comment]
Exactly, we are on the same page.

>  For multiple datasets, we need:
> 
> seed_i = (
>     base_seed
>     - worker_info.id
>     + torch.distributed.get_rank()
>     * (worker_info.num_workers * len(data.datasets))
>     + worker_info.id * len(data.datasets)
>     + i
> ) % (2 ** 32 - 1)

This seems to be equivalent to 
```                     
      seed_i = (
             base_seed
             + i * worker_info.num_workers
             + torch.distributed.get_rank() * len(data.datasets) * worker_info.num_workers
) % (2 ** 32 - 1)
```
but with the role of `i` and `worker_info.id` swapped because `base_seed` already has `worker_info.id` built in. Anyways, I will update the commit with the version you suggested since we have both verified that it works well.

**Comment by z-fabian** [comment]
@mmuckley Updated, it should be good now.

File: fastmri/pl_modules/data_module.py [hunk 3]
```diff
@@ -21,14 +21,40 @@ def worker_init_fn(worker_id):
...
+                    seed_i = (
+                        base_seed
+                        - worker_info.id
+                        + torch.distributed.get_rank()
+                        * (worker_info.num_workers * len(data.datasets))
+                        + worker_info.id * len(data.datasets)
+                        + i
+                    ) % (2 ** 32 - 1)
+                else:
+                    seed_i = (base_seed + i) % (2 ** 32 - 1)
```
Comment (mmuckley) [inline fastmri/pl_modules/data_module.py:49]: I think one more change is needed here. `seed_i = (base_seed + worker_info.id * len(data.datasets) + i) % (2 ** 32 -1)`.

File: fastmri/pl_modules/data_module.py [hunk 4]
```diff
@@ -21,14 +21,40 @@ def worker_init_fn(worker_id):
...
+
+    # Check if we are using DDP
+    is_ddp = False
+    if torch.distributed.is_available():
+        if torch.distributed.is_initialized():
+            is_ddp = True
+
     # for NumPy random seed we need it to be in this range
-    seed = worker_info.seed % (2 ** 32 - 1)  # pylint: disable=no-member
+    base_seed = worker_info.seed % (2 ** 32 - 1)  # pylint: disable=no-member
```
Comment (mmuckley) [inline fastmri/pl_modules/data_module.py:32]: Might be better to just do `% (2 ** 32 - 1)` at the `seed` function call.

**Review by mmuckley — CHANGES_REQUESTED** [review]
I think there's just one last change that's needed for the multiple-dataset, single-GPU case.

File: fastmri/pl_modules/data_module.py (hunk 3 above)
Comment (z-fabian) [reply fastmri/pl_modules/data_module.py:49]: Good point. Shouldn't it be `(base_seed - worker_info.id + worker_info.id * len(data.datasets) + i) % (2 ** 32 -1)`, since `base_seed` already has the worker id added to it internally?

File: fastmri/pl_modules/data_module.py (hunk 4 above)
Comment (z-fabian) [reply fastmri/pl_modules/data_module.py:32]: Sounds good. Do you mean only doing `% (2 ** 32 - 1)` at the very end at `dataset.transform.mask_func.rng.seed()` ?

Comment (mmuckley) [reply fastmri/pl_modules/data_module.py:32]: Yeah, it can be `dataset.transform.mask_func.rng.seed(seed % (2 ** 32 - 1))`, for example.

File: fastmri/pl_modules/data_module.py (hunk 3 above)
Comment (mmuckley) [reply fastmri/pl_modules/data_module.py:49]: You're right!

**Comment by z-fabian** [comment]
Updated! Mask seeds should be truly random now.

**Review by mmuckley — APPROVED** [review]
LGTM. Thanks for all this work!
